File: android/app/src/main/java/app/zkpassport/zkpassport/mrzscan/MRZScanConfig.kt
package app.zkpassport.zkpassport.mrzscan

import app.zkpassport.zkpassport.BuildConfig
import app.zkpassport.zkpassport.mrzscan.OpenCVImagePreprocessor

import android.os.Handler
import android.os.Looper
import android.util.Log

/**
 * Configuration for MRZ scanning with enhanced processing options
 */
object MRZScanConfig {
    
    // Feature flags
    var useEnhancedProcessor = true
    var enablePreprocessing = true
    var enableMultiFrame = true
    var enableAdvancedCorrection = true
    var enableMLKitAlternation = true  // Enable alternating between Tesseract and ML Kit
    var enableVisionMRZDetection = true  // Enable dynamic MRZ region detection using MLKit
    
    // Processing parameters
    var maxProcessingFrames = 30
    var minConsensusFrames = 1
    var consensusThreshold = 0.9f
    var minConfidence = 0.1f
    var frameExpiryMs = 5 * 60000L // 5 minutes
    
    // Image preprocessing - OPTIMIZED DEFAULTS (OpenCV)
    var preprocessingConfig = OpenCVImagePreprocessor.PreprocessingConfig(
        convertToGrayscale = true,
        binarize = true,                  // Adaptive threshold
        scaleToDPI = true,
        applyMorphology = true
    )
    
    // Character confusion matrix tuning
    var maxCorrectionCandidates = 200
    var enableContextAwareCorrection = true
    
    // Performance tuning
    var frameSkipCount = 1 
    var maxImageWidth = 1980  // Resize if larger (reduced from 1920 for better performance)
    var maxImageHeight = 1080   // Reduced from 1080 for better performance
    var minProcessingIntervalMs = 100L  // Minimum time between OCR operations (will be adjusted based on pool size if adaptive is enabled)
    var adaptiveFrameSkipping = true  // Dynamically adjust frame rate based on processing time
    
    // Parallel OCR processing (pool configuration)
    var enableParallelOCR = true  // Enable parallel OCR processing using multiple Tesseract instances
    var maxOcrPoolSize = 4  // Maximum number of Tesseract instances in the pool (limited by available memory)
    var minOcrPoolSize = 1  // Minimum number of Tesseract instances (even on low memory devices)
    var adaptiveProcessingInterval = true  // Automatically adjust minProcessingIntervalMs based on pool size
    
    // Base interval for adaptive calculation (used when pool size = 1)
    private const val BASE_PROCESSING_INTERVAL_MS = 50L
    
    // UI/UX settings
    var showProcessingStatus = true
    var hapticFeedbackEnabled = true
    
    // Camera settings
    var cameraResolutionWidth = 1980  // Camera resolution width (reduced from 1920 for better performance)
    var cameraResolutionHeight = 1080  // Camera resolution height (reduced from 1080 for better performance)
    var useHighestAvailableResolution = false // Disabled to prevent memory issues (was true)
    var maxCameraResolutionWidth = 1920  // Maximum camera resolution width (reduced from 4096)
    var maxCameraResolutionHeight = 1080 // Maximum camera resolution height (reduced from 3072)
    
    // Debug options
    var enableDebugLogging = BuildConfig.DEBUG
    var showDebugImageView = BuildConfig.DEBUG  // Show processed image in debug view
    var imageRotation = 90f  // Rotation angle for camera image (0, 90, 180, 270) - applied before OCR processing
    
    // Screen management
    var keepScreenAwake = true  // Keep screen awake during camera scanning to prevent sleep interruption
    
    // Haptic feedback
    var enableHapticFeedback = true  // Provide haptic feedback when valid MRZ frames are detected
    
    // MRZ text overlay
    var enableMRZOverlay = true      // Show real-time MRZ text overlay on camera preview
    var overlayTextSize = 10f        // Text size for MRZ overlay (in sp)
    var overlayTextColor = android.graphics.Color.GREEN  // Color for MRZ overlay text
    var showPlaceholderMRZ = true    // Show placeholder MRZ when no real MRZ is detected

    fun applyBinarizedHighConfidenceMode() {
        enableAdvancedCorrection = true
        frameSkipCount = 0
        useHighestAvailableResolution = false
        preprocessingConfig = OpenCVImagePreprocessor.PreprocessingConfig(
            convertToGrayscale = true,
            binarize = true,
            scaleToDPI = true,
            applyMorphology = true
        )

        // Only one frame is needed for high confidence
        maxProcessingFrames = 30
        minConsensusFrames = 1
        consensusThreshold = 0.9f
        minConfidence = 0.9f
    }

    fun applyBinarizedLowConfidenceMode() {
        enableAdvancedCorrection = true
        frameSkipCount = 0
        useHighestAvailableResolution = false
        preprocessingConfig = OpenCVImagePreprocessor.PreprocessingConfig(
            convertToGrayscale = true,
            binarize = true,
            scaleToDPI = true,
            applyMorphology = true
        )

        // Low confidence mode requires more frames to be processed
        maxProcessingFrames = 30
        minConsensusFrames = 8
        consensusThreshold = 0.1f
        minConfidence = 0.1f
    }
    
    /**
     * Calculate optimal processing interval based on pool size
     * More engines = shorter interval to keep them all busy
     */
    fun calculateOptimalInterval(poolSize: Int): Long {
        if (!adaptiveProcessingInterval || poolSize <= 0) {
            return BASE_PROCESSING_INTERVAL_MS
        }
        
        // Calculate interval inversely proportional to pool size
        val calculatedInterval = BASE_PROCESSING_INTERVAL_MS / poolSize
        
        // Ensure we don't go below a minimum threshold (50ms = 20 fps max)
        // to prevent overwhelming the system
        return calculatedInterval.coerceAtLeast(50L)
    }
    
    /**
     * Update processing interval based on actual pool size
     * Called after pool initialization
     */
    fun updateIntervalForPoolSize(poolSize: Int) {
        if (adaptiveProcessingInterval) {
            val oldInterval = minProcessingIntervalMs
            minProcessingIntervalMs = calculateOptimalInterval(poolSize)
            
            if (enableDebugLogging) {
                Log.i("MRZScanConfig", 
                    "Adaptive interval: ${oldInterval}ms -> ${minProcessingIntervalMs}ms for $poolSize engines " +
                    "(~${1000 / minProcessingIntervalMs} frames/sec)")
            }
        }
    }
}

File: ios/MrzScanner/MRZScanConfig.swift
//
//  MRZScanConfig.swift
//  Configuration for MRZ scanning with enhanced processing options
//  Based on Android's MRZScanConfig.kt
//

import Foundation
import CoreGraphics

struct MRZScanConfig {
    // MARK: - Feature Flags
    static var useEnhancedProcessor = true
    static var enablePreprocessing = true
    static var enableMultiFrame = true
    static var enableAdvancedCorrection = true
    static var enableVisionAlternation = true  // Enable alternating between Tesseract and Vision
    static var enableVisionMRZDetection = true  // Use Vision to detect MRZ bounding box before OCR
    
    // MARK: - Processing Parameters
    static var maxProcessingFrames = 30
    static var minConsensusFrames = 8
    static var consensusThreshold: Float = 0.9
    static var minConfidence: Float = 0.1
    static var frameExpiryMs: TimeInterval = 5 * 60 // 5 minutes
    
    // MARK: - Image Preprocessing Configuration
    struct PreprocessingConfig {
        var convertToGrayscale = true
        var binarize = true  // Adaptive threshold
        var scaleToDPI = false
        var applyMorphology = true
        
        // Target dimensions for MRZ processing
        static let targetMRZHeightTD3 = 200  // Passports (2-line)
        static let targetMRZHeightTD1 = 160  // ID cards (3-line)
        static let maxMRZHeight = 240        // Prevent extreme upscaling
    }
    
    static var preprocessingConfig = PreprocessingConfig()
    
    // MARK: - Character Confusion Matrix Tuning
    static var maxCorrectionCandidates = 200
    static var enableContextAwareCorrection = true
    
    // MARK: - Performance Tuning
    static var frameSkipCount = 1  // Skip every other frame (process every 2nd frame)
    static var maxImageWidth = 1980
    static var maxImageHeight = 1080
    static var minProcessingIntervalMs: TimeInterval = 0.15  // 150ms = ~6-7 fps for Tesseract
    static var tesseractProcessingIntervalMs: TimeInterval = 0.2  // 200ms = 5 fps for Tesseract-only
    static var visionProcessingIntervalMs: TimeInterval = 0.05  // 50ms = 20 fps for Vision-only
    static var adaptiveFrameSkipping = true
    
    // MARK: - Parallel OCR Processing
    static var enableParallelOCR = true
    static var maxOcrPoolSize = 4
    static var minOcrPoolSize = 1
    static var adaptiveProcessingInterval = true
    
    // Base interval for adaptive calculation (used when pool size = 1)
    private static let baseProcessingIntervalMs: TimeInterval = 0.05  // 50ms
    
    // MARK: - UI/UX Settings
    static var showProcessingStatus = true
    static var hapticFeedbackEnabled = true
    
    // MARK: - Camera Settings
    static var cameraResolutionWidth = 1980
    static var cameraResolutionHeight = 1080
    static var useHighestAvailableResolution = false
    static var maxCameraResolutionWidth = 1920
    static var maxCameraResolutionHeight = 1080
    
    // MARK: - Debug Options
    static var enableDebugLogging: Bool = {
        #if DEBUG
        return true
        #else
        return false
        #endif
    }()
    static var showDebugImageView: Bool = {
        #if DEBUG
        return true  // Show debug image view in debug builds
        #else
        return false
        #endif
    }()
    static var debugImageMaxSize: CGFloat = 200  // Max dimension for debug image
    static var debugImageScale: CGFloat = 0.25   // Max scale factor for debug image
    static var imageRotation: CGFloat = 90  // Rotation angle for camera image
    
    // MARK: - Screen Management
    static var keepScreenAwake = true
    
    // MARK: - Haptic Feedback
    static var enableHapticFeedback = true
    
    // MARK: - MRZ Text Overlay
    static var enableMRZOverlay = true
    static var overlayTextSize: CGFloat = 10.0
    static var overlayTextColor: UInt32 = 0x00FF00  // Green color (RGB hex)
    static var showPlaceholderMRZ = true
    
    /// Calculate optimal processing interval based on pool size
    static func calculateOptimalInterval(poolSize: Int) -> TimeInterval {
        guard adaptiveProcessingInterval, poolSize > 0 else {
            return baseProcessingIntervalMs
        }
        
        // Calculate interval inversely proportional to pool size
        let calculatedInterval = baseProcessingIntervalMs / Double(poolSize)
        
        // Ensure we don't go below a minimum threshold (50ms = 20 fps max)
        return max(calculatedInterval, 0.05)
    }
    
    /// Update processing interval based on actual pool size
    static func updateIntervalForPoolSize(_ poolSize: Int) {
        guard adaptiveProcessingInterval else { return }
        
        let oldInterval = minProcessingIntervalMs
        minProcessingIntervalMs = calculateOptimalInterval(poolSize: poolSize)
        
        if enableDebugLogging {
            print("MRZScanConfig: Adaptive interval: \(oldInterval)s -> \(minProcessingIntervalMs)s for \(poolSize) engines (~\(Int(1.0 / minProcessingIntervalMs)) frames/sec)")
        }
    }
}
